tss2-dlopen/tss2-dlopen-esys.c

```diff
@@ -0,0 +1,2199 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*******************************************************************************
+ * Copyright 2021, Fraunhofer SIT
+ * All rights reserved.
+ *******************************************************************************/
+
+/**
+ * The purpose of this file is to copy it into your project and
+ * include it during compilation if you don't want to link against
+ * libtss2-esys at compile time.
+ * It will attempt to load libtss2-esys.so during runtime.
+ * It will either work similarly to directly linking to libtss2-esys.so
+ * at compile-time or return a NOT_IMPLEMENTED error.
+ *
+ * For new versions of this file, please check:
+ * http://github.com/tpm2-software/tpm2-tss/tss2-dlopen
+*/
+
+#include <dlfcn.h>
+#include <stdio.h>
+#include <tss2/tss2_esys.h>
+
+#define str(s) xstr(s)
+#define xstr(s) #s
+
+#ifdef ENABLE_WARN
+#define WARN(str, ...) do { fprintf(stderr, "WARNING: " str "\n", ## __VA_ARGS__); } while (0)
+#else /* ENABLE_WARN */
+#define WARN(...) do { ; } while (0)
+#endif /* ENABLE_WARN */
+
+static void *esys_dlhandle = NULL;
+
+TSS2_RC
+Esys_Initialize(
+    ESYS_CONTEXT **esys_context,
+    TSS2_TCTI_CONTEXT *tcti,
+    TSS2_ABI_VERSION *abiVersion)
+{
+    if (!esys_dlhandle) {
+        esys_dlhandle = dlopen("tss2-esys", RTLD_NOW | RTLD_LOCAL);
+        if (!esys_dlhandle) {
+            WARN("Library libtss2-esys not found: %s.", dlerror());
+            return TSS2_ESYS_RC_NOT_IMPLEMENTED;
+        }
+    }
+
+    static TSS2_RC (*sym) (ESYS_CONTEXT **esys_context, TSS2_TCTI_CONTEXT *tcti, TSS2_ABI_VERSION *abiVersion) = NULL;
+    if (!sym)
+        sym = dlsym(esys_dlhandle, "Esys_Initialize");
+    if (!sym) {
+        WARN("Function Esys_Initialize not found.");
+        return TSS2_ESYS_RC_NOT_IMPLEMENTED;
+    }
```

Comment:
personally, I would do all the heavy lifting in `Esys_Initialize()` and do something like:
```C
malloc(esys_hdr + sizeof(function table);
```
So your structure for esys ctx pointer is:
```
Normal esys ctx
function table
```

Then in your subsequent calls no dlsym needed, you just:
```
if (esys_ctx) {
   esys_ctx->fn1()
}
```
However that would mean esys ctx would be incompatible between dlopen and standard versions, but I don't know how they would end up with both symbol names in their program without doing something weird to bring it on themselves. You could pass a dl-open ctx to the lib version but not vice versa.

Reply:
I guess you could just make the whole fn table static global and initialize it in Esys_Initialize, but you would want to make it thread safe, which it currently isn't with your use of static, but the dlopen/dlsym/dlclose are MT safe. However, the scenario of N>1 threads updating the sym variable which are not atomic.

Reply:
Is it really worth moving all the dlsym() calls into the intiialize ?
It will end up being slower because we make a lot more dlsym() calls than the application really needs. So I see no benefit here.

Wrt being thread safe: The pointer writes are atomic and calling dlsym twice will also not be a problem. If we end up with two handles to libesys / etc it's also ok, because the libtss2_* do not have global state so we can even mix and mingle between those.
Thus, I'd keep the current lightweight version.
If people want to do it differently when they copy these files, they can of course rewrite them to their liking.

Reply:
Its MT-Safe so we're fine here.